### PR TITLE
Add Relational GIN implementation

### PR DESCRIPTION
**Pull request by mmjb**
**[PR under construction]**

This is an implementation of a relational GIN variant, see `gnns/rgin.py` for a more precise description. I've verified that the collapse onto a single edge type works and has comparable results as the original GIN paper, even though I stripped out all of their batchnorms (including inside the MLPs) and replaced them by a layernorm at the end of propagation. I'll do the remaining hyperparameter search for QM9/VarMisuse over the weekend, I think, but early feedback would be helpful.

### CODE REVIEW

File: gnns/rgin.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            edge_type_to_edge_mlp.append(
+                MLP(out_size=state_dim,
+                    hidden_layers=num_MLP_hidden_layers,
+                    activation_fun=activation_fn))
+        edge_type_to_message_targets.append(adjacency_list_for_edge_type[:, 1])
+    # Initialize epsilon: Note that we merge the 1 + \epsilon from the Def. above:
+    if learn_epsilon:
+        epsilon = tf.get_variable("epsilon", shape=(), dtype=tf.float32, initializer=tf.ones_initializer, trainable=True)
+    else:
+        epsilon = 1
```
Comment (mallamanis) [inline gnns/rgin.py]: Minor technical thing: maybe allow the user to pass in `epsilon: Optional[float]=None` and when `epsilon is None` imply that it's learnable?

Also, as you note there is a discrepancy here, `epsilon = 1 + ε`  (of Eq 4.1). What about renaming `epsilon` to `epsilon_plus_one` (or something like that)?

File: utils/utils.py [hunk 2]
```diff
@@ -76,35 +75,45 @@ def micro_f1(logits, labels):
...
+            out_size: Dimensionality of output.
+            hidden_layers: Either an integer determining number of hidden layers, who will have out_size units each;
+                or list of integers whose lengths determines the number of hidden layers and whose contents the
+                number of units in each layer.
+            use_biases: Flag indicating use of bias in fully connected layers.
+            activation_fun: Activation function applied between hidden layers (NB: the output of the MLP
+                is always the direct result of a linear transformation)
+            dropout_rate: Dropout applied to inputs of each MLP layer.
+        """
+        self.__dropout_rate = dropout_rate
```
Comment (mallamanis) [inline utils/utils.py]: I would be tempted to add a 
```python
if hidden_layers > 1:
    assert activation_fun is not None, 'Multiple linear layers without an activation`
```
I suspect that in some rare cases a low-rank approximation can use this, but it should be rare?

File: utils/utils.py [hunk 3]
```diff
@@ -76,35 +75,45 @@ def micro_f1(logits, labels):
...
+        if isinstance(hidden_layers, int):
+            hidden_layer_sizes = [out_size] * hidden_layers
+        else:
+            hidden_layer_sizes = hidden_layers
+
+        self.__layers = []  # type: List[tf.layers.Dense]
+        for _ in hidden_layer_sizes:
+            self.__layers.append(tf.layers.Dense(units=out_size,
+                                                 use_bias=use_biases,
+                                                 activation=activation_fun))
```
Comment (mallamanis) [inline utils/utils.py]: `units=out_size` is wrong, you probably want to use the elements in `hidden_layer_sizes`

File: utils/utils.py [hunk 4]
```diff
@@ -76,35 +75,45 @@ def micro_f1(logits, labels):
...
+                 activation_fun: Optional[Callable[[tf.Tensor], tf.Tensor]] = tf.nn.relu,
+                 dropout_rate: Union[float, tf.Tensor] = 0.0):
+        """
+        Create new MLP with given number of hidden layers.
+
+        Arguments:
+            out_size: Dimensionality of output.
+            hidden_layers: Either an integer determining number of hidden layers, who will have out_size units each;
+                or list of integers whose lengths determines the number of hidden layers and whose contents the
+                number of units in each layer.
```
Comment (mallamanis) [inline utils/utils.py:93]: The `out_size` and `hidden_layers` seem to be two conflicting set of parameters in some cases. Maybe remove the out_size in favor of the `hidden_layers`?

**Review by mallamanis — COMMENTED** [review]
I don't have any substantial comment here, beyond some coding suggestions...
